[FX] Delete values after their last use

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48631 [FX] Delete values after their last use**

This deletes values after they're done being used in the generated Python code so that memory usage is optimal

Resnet18 Example:

Before

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    bn1 = self.bn1(conv1)
    relu_1 = self.relu(bn1)
    maxpool = self.maxpool(relu_1)
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    avgpool = self.avgpool(layer4_2)
    flatten_1 = torch.flatten(avgpool, 1)
    fc = self.fc(flatten_1)
    return fc

```

After

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    x = None
    bn1 = self.bn1(conv1)
    conv1 = None
    relu_1 = self.relu(bn1)
    bn1 = None
    maxpool = self.maxpool(relu_1)
    relu_1 = None
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_conv1 = None
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_bn1 = None
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_relu = None
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    layer1_0_conv2 = None
    add_1 = layer1_0_bn2 + maxpool
    maxpool = layer1_0_bn2 = None
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    add_1 = None
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_conv1 = None
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_bn1 = None
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_relu = None
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    layer1_1_conv2 = None
    add_2 = layer1_1_bn2 + layer1_1
    layer1_1_bn2 = layer1_1 = None
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    add_2 = None
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_conv1 = None
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_bn1 = None
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_relu = None
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_conv2 = None
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer1_2 = None
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    layer2_0_downsample_0 = None
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_0_downsample_1 = layer2_0_bn2 = None
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    add_3 = None
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_conv1 = None
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_bn1 = None
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_relu = None
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    layer2_1_conv2 = None
    add_4 = layer2_1_bn2 + layer2_1
    layer2_1 = layer2_1_bn2 = None
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    add_4 = None
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_conv1 = None
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_bn1 = None
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_relu = None
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_conv2 = None
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer2_2 = None
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    layer3_0_downsample_0 = None
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_0_bn2 = layer3_0_downsample_1 = None
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    add_5 = None
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_conv1 = None
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_bn1 = None
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_relu = None
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    layer3_1_conv2 = None
    add_6 = layer3_1_bn2 + layer3_1
    layer3_1_bn2 = layer3_1 = None
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    add_6 = None
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_conv1 = None
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_bn1 = None
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_relu = None
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_conv2 = None
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer3_2 = None
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    layer4_0_downsample_0 = None
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_0_bn2 = layer4_0_downsample_1 = None
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    add_7 = None
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_conv1 = None
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_bn1 = None
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_relu = None
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    layer4_1_conv2 = None
    add_8 = layer4_1_bn2 + layer4_1
    layer4_1_bn2 = layer4_1 = None
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    add_8 = None
    avgpool = self.avgpool(layer4_2)
    layer4_2 = None
    flatten_1 = torch.flatten(avgpool, 1)
    avgpool = None
    fc = self.fc(flatten_1)
    flatten_1 = None
    return fc
```


Differential Revision: [D25235981](https://our.internmc.facebook.com/intern/diff/D25235981)

Commit:
[FX] Delete values after their last use

[ghstack-poisoned]

Commit:
Update on "[FX] Delete values after their last use"




[ghstack-poisoned]

Commit:
Update on "[FX] Delete values after their last use"


This deletes values after they're done being used in the generated Python code.

Resnet18 Example:

Before

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    bn1 = self.bn1(conv1)
    relu_1 = self.relu(bn1)
    maxpool = self.maxpool(relu_1)
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    avgpool = self.avgpool(layer4_2)
    flatten_1 = torch.flatten(avgpool, 1)
    fc = self.fc(flatten_1)
    return fc

```

After

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    del x
    bn1 = self.bn1(conv1)
    del conv1
    relu_1 = self.relu(bn1)
    del bn1
    maxpool = self.maxpool(relu_1)
    del relu_1
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    del layer1_0_conv1
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    del layer1_0_bn1
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    del layer1_0_relu
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    del layer1_0_conv2
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    del add_1
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    del layer1_1_conv1
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    del layer1_1_bn1
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    del layer1_1_relu
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    del layer1_1_conv2
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    del add_2
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    del layer2_0_conv1
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    del layer2_0_bn1
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    del layer2_0_relu
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    del layer2_0_conv2
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    del layer1_2
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    del layer2_0_downsample_0
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    del add_3
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    del layer2_1_conv1
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    del layer2_1_bn1
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    del layer2_1_relu
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    del layer2_1_conv2
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    del add_4
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    del layer3_0_conv1
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    del layer3_0_bn1
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    del layer3_0_relu
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    del layer3_0_conv2
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    del layer2_2
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    del layer3_0_downsample_0
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    del add_5
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    del layer3_1_conv1
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    del layer3_1_bn1
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    del layer3_1_relu
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    del layer3_1_conv2
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    del add_6
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    del layer4_0_conv1
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    del layer4_0_bn1
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    del layer4_0_relu
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    del layer4_0_conv2
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    del layer3_2
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    del layer4_0_downsample_0
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    del add_7
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    del layer4_1_conv1
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    del layer4_1_bn1
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    del layer4_1_relu
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    del layer4_1_conv2
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    del add_8
    avgpool = self.avgpool(layer4_2)
    del layer4_2
    flatten_1 = torch.flatten(avgpool, 1)
    del avgpool
    fc = self.fc(flatten_1)
    del flatten_1
    return fc

```


Differential Revision: [D25235981](https://our.internmc.facebook.com/intern/diff/D25235981)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -487,6 +487,30 @@ def type_repr(o : Any):
+        # tracking and deleting values.
+        seen_nodes : Set[Node] = set()
+
+        def delete_unused_values(node : Node):
```
The cheaper way to do this is to find the first use of the node in the reverse node list. This prevents having to do the subset checks which is u^2 where u is the node with the maximum number of uses.

Commit:
Update on "[FX] Delete values after their last use"


This deletes values after they're done being used in the generated Python code so that memory usage is optimal

Resnet18 Example:

Before

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    bn1 = self.bn1(conv1)
    relu_1 = self.relu(bn1)
    maxpool = self.maxpool(relu_1)
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    avgpool = self.avgpool(layer4_2)
    flatten_1 = torch.flatten(avgpool, 1)
    fc = self.fc(flatten_1)
    return fc

```

After

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    x = None
    bn1 = self.bn1(conv1)
    conv1 = None
    relu_1 = self.relu(bn1)
    bn1 = None
    maxpool = self.maxpool(relu_1)
    relu_1 = None
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_conv1 = None
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_bn1 = None
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_relu = None
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    layer1_0_conv2 = None
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    add_1 = None
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_conv1 = None
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_bn1 = None
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_relu = None
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    layer1_1_conv2 = None
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    add_2 = None
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_conv1 = None
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_bn1 = None
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_relu = None
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_conv2 = None
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer1_2 = None
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    layer2_0_downsample_0 = None
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    add_3 = None
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_conv1 = None
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_bn1 = None
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_relu = None
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    layer2_1_conv2 = None
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    add_4 = None
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_conv1 = None
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_bn1 = None
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_relu = None
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_conv2 = None
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer2_2 = None
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    layer3_0_downsample_0 = None
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    add_5 = None
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_conv1 = None
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_bn1 = None
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_relu = None
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    layer3_1_conv2 = None
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    add_6 = None
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_conv1 = None
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_bn1 = None
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_relu = None
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_conv2 = None
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer3_2 = None
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    layer4_0_downsample_0 = None
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    add_7 = None
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_conv1 = None
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_bn1 = None
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_relu = None
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    layer4_1_conv2 = None
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    add_8 = None
    avgpool = self.avgpool(layer4_2)
    layer4_2 = None
    flatten_1 = torch.flatten(avgpool, 1)
    avgpool = None
    fc = self.fc(flatten_1)
    flatten_1 = None
    return fc

```


Differential Revision: [D25235981](https://our.internmc.facebook.com/intern/diff/D25235981)

[ghstack-poisoned]

Commit:
Update on "[FX] Delete values after their last use"


This deletes values after they're done being used in the generated Python code so that memory usage is optimal

Resnet18 Example:

Before

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    bn1 = self.bn1(conv1)
    relu_1 = self.relu(bn1)
    maxpool = self.maxpool(relu_1)
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    avgpool = self.avgpool(layer4_2)
    flatten_1 = torch.flatten(avgpool, 1)
    fc = self.fc(flatten_1)
    return fc

```

After

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    x = None
    bn1 = self.bn1(conv1)
    conv1 = None
    relu_1 = self.relu(bn1)
    bn1 = None
    maxpool = self.maxpool(relu_1)
    relu_1 = None
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_conv1 = None
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_bn1 = None
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_relu = None
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    layer1_0_conv2 = None
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    add_1 = None
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_conv1 = None
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_bn1 = None
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_relu = None
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    layer1_1_conv2 = None
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    add_2 = None
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_conv1 = None
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_bn1 = None
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_relu = None
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_conv2 = None
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer1_2 = None
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    layer2_0_downsample_0 = None
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    add_3 = None
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_conv1 = None
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_bn1 = None
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_relu = None
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    layer2_1_conv2 = None
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    add_4 = None
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_conv1 = None
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_bn1 = None
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_relu = None
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_conv2 = None
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer2_2 = None
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    layer3_0_downsample_0 = None
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    add_5 = None
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_conv1 = None
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_bn1 = None
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_relu = None
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    layer3_1_conv2 = None
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    add_6 = None
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_conv1 = None
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_bn1 = None
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_relu = None
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_conv2 = None
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer3_2 = None
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    layer4_0_downsample_0 = None
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    add_7 = None
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_conv1 = None
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_bn1 = None
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_relu = None
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    layer4_1_conv2 = None
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    add_8 = None
    avgpool = self.avgpool(layer4_2)
    layer4_2 = None
    flatten_1 = torch.flatten(avgpool, 1)
    avgpool = None
    fc = self.fc(flatten_1)
    flatten_1 = None
    return fc
```


Differential Revision: [D25235981](https://our.internmc.facebook.com/intern/diff/D25235981)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -487,6 +487,41 @@ def type_repr(o : Any):
+            not used in the remainder of the code are freed and the memory usage
+            of the code is optimal.
+            """
+            nodes_to_delete : Set[Node] = set()
```
sets are unordered, so this creates non-determinisitic code. making this a dictionary instead would make it deterministic.

zdevito replied:
it's also possible to just construct this set in the `register_last_uses` function by maintaining a user -> last_uses map as it goes andd appending nodes as they are found. Its minor but it avoids having to scan the argument list of a node twice.

zdevito commented on torch/fx/graph.py:
```diff
@@ -502,13 +537,15 @@ def type_repr(o : Any):
                 body.append(
                     f'{node.name} = {_format_target(repr(node.args[0]), node.target)}'
                     f'({_format_args(node.args[1:], node.kwargs)})\n')
+                delete_unused_values(node)
```
the control structure here is confusing. I'd expect delete_unused_values to only get issued once, but I see it inside nested if statements. Maybe refactor `for node in self.nodes: emitNode(node); emit_delete_unused_values(node);`

zdevito reviewed (approved):
I have some minor structural things to address but looks right overall.

Commit:
Update on "[FX] Delete values after their last use"


This deletes values after they're done being used in the generated Python code so that memory usage is optimal

Resnet18 Example:

Before

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    bn1 = self.bn1(conv1)
    relu_1 = self.relu(bn1)
    maxpool = self.maxpool(relu_1)
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    add_1 = layer1_0_bn2 + maxpool
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    add_2 = layer1_1_bn2 + layer1_1
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    add_4 = layer2_1_bn2 + layer2_1
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    add_6 = layer3_1_bn2 + layer3_1
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    add_8 = layer4_1_bn2 + layer4_1
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    avgpool = self.avgpool(layer4_2)
    flatten_1 = torch.flatten(avgpool, 1)
    fc = self.fc(flatten_1)
    return fc

```

After

```
import torch
def forward(self, x):
    conv1 = self.conv1(x)
    x = None
    bn1 = self.bn1(conv1)
    conv1 = None
    relu_1 = self.relu(bn1)
    bn1 = None
    maxpool = self.maxpool(relu_1)
    relu_1 = None
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1)
    layer1_0_conv1 = None
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1)
    layer1_0_bn1 = None
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu)
    layer1_0_relu = None
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2)
    layer1_0_conv2 = None
    add_1 = layer1_0_bn2 + maxpool
    maxpool = layer1_0_bn2 = None
    layer1_1 = getattr(self.layer1, "0").relu(add_1)
    add_1 = None
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1)
    layer1_1_conv1 = None
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1)
    layer1_1_bn1 = None
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu)
    layer1_1_relu = None
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2)
    layer1_1_conv2 = None
    add_2 = layer1_1_bn2 + layer1_1
    layer1_1_bn2 = layer1_1 = None
    layer1_2 = getattr(self.layer1, "1").relu(add_2)
    add_2 = None
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_2)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1)
    layer2_0_conv1 = None
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1)
    layer2_0_bn1 = None
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu)
    layer2_0_relu = None
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2)
    layer2_0_conv2 = None
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_2)
    layer1_2 = None
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0)
    layer2_0_downsample_0 = None
    add_3 = layer2_0_bn2 + layer2_0_downsample_1
    layer2_0_downsample_1 = layer2_0_bn2 = None
    layer2_1 = getattr(self.layer2, "0").relu(add_3)
    add_3 = None
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1)
    layer2_1_conv1 = None
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1)
    layer2_1_bn1 = None
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu)
    layer2_1_relu = None
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2)
    layer2_1_conv2 = None
    add_4 = layer2_1_bn2 + layer2_1
    layer2_1 = layer2_1_bn2 = None
    layer2_2 = getattr(self.layer2, "1").relu(add_4)
    add_4 = None
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_2)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1)
    layer3_0_conv1 = None
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1)
    layer3_0_bn1 = None
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu)
    layer3_0_relu = None
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2)
    layer3_0_conv2 = None
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_2)
    layer2_2 = None
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0)
    layer3_0_downsample_0 = None
    add_5 = layer3_0_bn2 + layer3_0_downsample_1
    layer3_0_bn2 = layer3_0_downsample_1 = None
    layer3_1 = getattr(self.layer3, "0").relu(add_5)
    add_5 = None
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1)
    layer3_1_conv1 = None
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1)
    layer3_1_bn1 = None
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu)
    layer3_1_relu = None
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2)
    layer3_1_conv2 = None
    add_6 = layer3_1_bn2 + layer3_1
    layer3_1_bn2 = layer3_1 = None
    layer3_2 = getattr(self.layer3, "1").relu(add_6)
    add_6 = None
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_2)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1)
    layer4_0_conv1 = None
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1)
    layer4_0_bn1 = None
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu)
    layer4_0_relu = None
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2)
    layer4_0_conv2 = None
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_2)
    layer3_2 = None
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0)
    layer4_0_downsample_0 = None
    add_7 = layer4_0_bn2 + layer4_0_downsample_1
    layer4_0_bn2 = layer4_0_downsample_1 = None
    layer4_1 = getattr(self.layer4, "0").relu(add_7)
    add_7 = None
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1)
    layer4_1_conv1 = None
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1)
    layer4_1_bn1 = None
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu)
    layer4_1_relu = None
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2)
    layer4_1_conv2 = None
    add_8 = layer4_1_bn2 + layer4_1
    layer4_1_bn2 = layer4_1 = None
    layer4_2 = getattr(self.layer4, "1").relu(add_8)
    add_8 = None
    avgpool = self.avgpool(layer4_2)
    layer4_2 = None
    flatten_1 = torch.flatten(avgpool, 1)
    avgpool = None
    fc = self.fc(flatten_1)
    flatten_1 = None
    return fc
```


Differential Revision: [D25235981](https://our.internmc.facebook.com/intern/diff/D25235981)

[ghstack-poisoned]
